src/Enzo/enzo_EnzoConstrainedTransport.cpp

```diff
@@ -10,30 +10,289 @@
+  // interfaces for that dimension (they exclude exterior faces of the block)
+  weight_l_[0] = EFlt3DArray(  mz,   my, mx-1);
+  weight_l_[1] = EFlt3DArray(  mz, my-1,   mx);
+  weight_l_[2] = EFlt3DArray(mz-1,   my,   mx);
+
+  // allocate arrays to hold for each component of the edge-centered electric
+  // fields. The array for component i is cell-centered along i and face
+  // centered along j and k.
+  edge_efield_l_[0] = EFlt3DArray(mz-1, my-1,   mx);
+  edge_efield_l_[1] = EFlt3DArray(mz-1,   my, mx-1);
+  edge_efield_l_[2] = EFlt3DArray(  mz, my-1, mx-1);
+
+  // allocate array to temporarily store cell-centered e-field
+  center_efield_ = EFlt3DArray(mz,my,mx);
+}
+
+//----------------------------------------------------------------------
+
+void EnzoConstrainedTransport::check_required_fields()
+{
+  FieldDescr * field_descr = cello::field_descr();
+  ASSERT("EnzoConstrainedTransport::update_refresh",
+	 ("There must be face-centered permanent fields called \"bfieldi_x\","
+	  "\"bfield_y\" and \"bfield_z\"."),
+	 (field_descr->is_field("bfield_x") &&
+	  field_descr->is_field("bfield_y") &&
+	  field_descr->is_field("bfield_z")));
+
+  std::vector<std::string> names = {"bfieldi_x", "bfieldi_y", "bfieldi_z"};
+  std::vector<std::string> axes = {"x", "y", "z"};
+  for (std::size_t i = 0; i < 3; i++){
+    std::string name = names[i];
+    // first check that field exists
+    ASSERT1("EnzoConstrainedTransport::update_refresh",
+	    "There must be face-centered permanent fields called \"%s\"",
+	    name.c_str(), field_descr->is_field(name));
+
+    int field_id = field_descr->field_id(name);
+    // next check the centering of the field
+    int centering[3] = {0, 0, 0};
+    field_descr->centering(field_id, &centering[0], &centering[1],
+			   &centering[2]);
+    for (std::size_t j = 0; j<3; j++){
+      if (j!=i){
+	ASSERT2("EnzoConstrainedTransport::update_refresh",
+		"The \"%s\" field must be cell-centered along the %s-axis",
+		name.c_str(), axes[j].c_str(), centering[j] == 0);
+      } else {
+	ASSERT2("EnzoConstrainedTransport::update_refresh",
+		"The \"%s\" field must be face-centered along the %s-axis",
+		name.c_str(), axes[j].c_str(), centering[j] == 1);
+      }
+    }
+
+  }
+}
+
+//----------------------------------------------------------------------
+
+void EnzoConstrainedTransport::increment_partial_timestep() throw()
+{
+  partial_timestep_index_++;
+  ASSERT2("EnzoConstrainedTransport::increment_partial_timestep",
+	  ("This should not be called more than %d times to transition "
+	   "between the %d partial timestep(s)"),
+	  num_partial_timesteps_-1, num_partial_timesteps_,
+	  num_partial_timesteps_ > partial_timestep_index_);
+}
+
+//----------------------------------------------------------------------
+
+void EnzoConstrainedTransport::correct_reconstructed_bfield
+(EnzoEFltArrayMap &l_map, EnzoEFltArrayMap &r_map, int dim, int stale_depth)
```

Comment:
```suggestion
(const EnzoEFltArrayMap &l_map, const EnzoEFltArrayMap &r_map, int dim, int stale_depth)
```

Reply:
I think I originally was holding off on this because I wasn't sure about the desire behavior for a `const EnzoEFltArrayMap`... Do we want it to hold:
- `const CelloArray<enzo_float, 3>` where values within the array are still mutable
- `const CelloArray<const enzo_float, 3>`. Strictly speaking, this isn't supported by CelloArray, but I've adopted the Kokkos documentation. The idea is that the contents are read-only (going with this approach would require that we add this feature to the CelloArray.

I'm still not sure which approach we want to go with. I would appreciate feedback on this point.

You've pointed out other places where this distinction is moot and I should make the references const, either way.